waspls/src/Wasp/LSP/Handlers.hs

```diff
@@ -0,0 +1,131 @@
+module Wasp.LSP.Handlers
+  ( initializedHandler,
+    didOpenHandler,
+    didChangeHandler,
+    didSaveHandler,
+  )
+where
+
+import Control.Lens ((+~), (^.))
+import Control.Monad.Trans (lift)
+import Control.Monad.Trans.Except (throwE)
+import Data.Function ((&))
+import Data.Text (Text)
+import qualified Data.Text as T
+import Language.LSP.Server (Handlers, LspT)
+import qualified Language.LSP.Server as LSP
+import qualified Language.LSP.Types as LSP
+import qualified Language.LSP.Types.Lens as LSP
+import Language.LSP.VFS (virtualFileText)
+import qualified Wasp.Analyzer
+import qualified Wasp.Analyzer.AnalyzeError as WE
+import Wasp.Analyzer.Parser (Ctx (Ctx))
+import Wasp.Analyzer.Parser.SourceRegion (getRgnEnd, getRgnStart)
+import Wasp.LSP.Core (ServerConfig, ServerError (ServerError), ServerM, Severity (..))
+
+-- LSP notification and request handlers
+
+-- | "Initialized" notification is sent when the client is started. We don't
+-- have anything we need to do at initialization, but this is required to be
+-- implemented.
+--
+-- The client starts the LSP at its own discretion, but commonly this is done
+-- either when:
+--
+-- - A file of the associated language is opened (in this case `.wasp`)
+-- - A workspace is opened that has a project structure associated with the
+--   language (in this case, a `main.wasp` file in the root folder of the
+--   workspace)
+initializedHandler :: Handlers ServerM
+initializedHandler =
+  LSP.notificationHandler LSP.SInitialized $ const (return ())
+
+-- | "TextDocumentDidOpen" is sent by the client when a new document is opened.
+-- `diagnoseWaspFile` is run to analyze the newly opened document.
+didOpenHandler :: Handlers ServerM
+didOpenHandler =
+  LSP.notificationHandler LSP.STextDocumentDidOpen $ diagnoseWaspFile . extractUri
+
+-- | "TextDocumentDidChange" is sent by the client when a document is changed
+-- (i.e. when the user types/deletes text). `diagnoseWaspFile` is run to
+-- analyze the changed document.
+didChangeHandler :: Handlers ServerM
+didChangeHandler =
+  LSP.notificationHandler LSP.STextDocumentDidChange $ diagnoseWaspFile . extractUri
+
+-- | "TextDocumentDidSave" is sent by the client when a document is saved.
+-- `diagnoseWaspFile` is run to analyze the new contents of the document.
+didSaveHandler :: Handlers ServerM
+didSaveHandler =
+  LSP.notificationHandler LSP.STextDocumentDidSave $ diagnoseWaspFile . extractUri
```

Comment:
I think it would, but I felt it's probably better to be paranoid about reprocessing the file to make sure it doesn't get out of sync. At least until we start seeing performance issues, I think we should run on save anyways and possibly duplicate a little bit of work.

Reply:
Ok makes sense, yeah I agree it is probably better to go over vs under to start and we can monitor from there. 👍🏻